feat(download) support external location picker

AckslD opened the pull request:
Closes #1365 

I haven't added tests yes cause it would be great to have some feedback if others think this is a good way to handle things.

What this MR enables is a new command and keybind in the download prompt to trigger an external folder picker. When the picker finishes the prompt gets updated with a new default value.

One could do some things differently here. For example:
* When a folder is chosen, directly accept the prompt. However I thought it would be better to first see what was chosen and then accept.
* Allow a setting such that the external picker is triggered directly. But is seems nice to still have the option to for example choose to open the download etc.

The-Compiler commented on qutebrowser/mainwindow/prompt.py:
```diff
@@ -467,6 +467,27 @@ def prompt_yank(self, sel=False):
+        """Choose a location using a configured external picker."""
+        picker = config.val.fileselect.folder.command
+        if not picker:
+            log.webview.warning(
```
```suggestion
            log.prompt.warning(
```

would probably be more appropriate - but I suppose this should actually be a

```suggestion
            raise cmdutils.CommandError(
```

because the command cannot continue, and this should be communicated to the user (not just logged to the terminal, which is often missing or ignored.

agenbite commented:
I tried this and it works as advertised. I used ranger, and had to add this to my config (I was confused at the beginning about which option should be set): 
```
config.set("fileselect.folder.command", ['xfce4-terminal', '--disable-server', '-T', 'auxiliary file select', "--geometry", "86x24+40+60", '-e', 'ranger --show-only-dirs --choosedir={}'])
```

With ranger, it's a bit weird that you need to quit in order to accept the selected folder. But you get used to it.

With respect to the alternatives mentioned above, I'm fine with accepting the selected folder in two steps. And yes it'd be nice to have a setting to trigger the external picker directly, but it's not a blocker, imo.

It'd be great to have this merged. Thanks!

The-Compiler commented on qutebrowser/mainwindow/prompt.py:
```diff
@@ -467,6 +467,27 @@ def prompt_yank(self, sel=False):
+        question = self._prompt.question
+        # choose_file already checks that this is max one folder
+        question.default = folders[0]
+        self._prompt._lineedit.setText(question.default)
```
This should instead be a method in `_BasePrompt` raising `UnsupportedOperationError`, which is then overridden in `LineEditPrompt`. Right now I suspect that this would crash with an `AttributeError` if this command is run with e.g. an authentication or yes/no prompt (JS confirm, TLS error, etc.).

I think the method can be kept general (i.e. `set_text` rather than `set_path` or somesuch), which would also allow implementing an equivalent to `:set-cmd-text` for prompt mode later.

Alternatively, we could just call `accept` here as you mention, I'll answer to that in my main comment.

AckslD replied:
Changed to directly accept in 65c3007fc07f708e716c9cb14fefd10bf5f240d7, is that as simple as calling `self._prompt.accept` with the path to the folder?

The-Compiler commented on qutebrowser/mainwindow/prompt.py:
```diff
@@ -467,6 +467,27 @@ def prompt_yank(self, sel=False):
+        instance='prompt-container', scope='window',
+        modes=[usertypes.KeyMode.prompt])
+    def prompt_external_picker(self):
+        """Choose a location using a configured external picker."""
```
Should we abort early if the prompt isn't a `FilenamePrompt`? While the code would in theory work when e.g. being asked for the name of a quickmark, that really doesn't make any sense.

AckslD replied:
Done in 65c3007fc07f708e716c9cb14fefd10bf5f240d7

The-Compiler commented on qutebrowser/mainwindow/prompt.py:
```diff
@@ -467,6 +467,27 @@ def prompt_yank(self, sel=False):
+            return
+        assert self._prompt is not None
+        question = self._prompt.question
+        # choose_file already checks that this is max one folder
```
Maybe add an assert for it here?

AckslD replied:
Done in 65c3007fc07f708e716c9cb14fefd10bf5f240d7

The-Compiler commented on qutebrowser/mainwindow/prompt.py:
```diff
@@ -821,6 +842,7 @@ def _allowed_commands(self):
             ('prompt-open-download', "Open download"),
             ('prompt-open-download --pdfjs', "Open download via PDF.js"),
             ('prompt-yank', "Yank URL"),
+            ('prompt-external-picker', "External location picker"),
```
```suggestion
            ('prompt-external-picker', "Launch external location picker"),
```

for consistency with the other descriptions.

AckslD replied:
Done in 65c3007fc07f708e716c9cb14fefd10bf5f240d7

The-Compiler commented on qutebrowser/mainwindow/prompt.py:
```diff
@@ -467,6 +467,27 @@ def prompt_yank(self, sel=False):
+    @cmdutils.register(
+        instance='prompt-container', scope='window',
+        modes=[usertypes.KeyMode.prompt])
+    def prompt_external_picker(self):
```
We call this `fileselect` so far, e.g. in the settings. Should the command be named something like `prompt_fileselect_external` or so too?

AckslD replied:
Done in 65c3007fc07f708e716c9cb14fefd10bf5f240d7

The-Compiler reviewed (changes requested):
Thanks and sorry for the delay! I added a couple of comments.

I agree with the approach of keeping the existing prompt (so files can e.g. be opened, or easily confirmed if one e.g. just wants the default download folder) and having a command to spawn the file selector from it. It's very similar to what we have for e.g. PDF.js too.

As for your thoughts:

> * When a folder is chosen, directly accept the prompt. However I thought it would be better to first see what was chosen and then accept.

I'm not sure, I think I would prefer directly accepting the prompt. It will make the code a bit simpler (see my other comment on that), and I don't see the benefit of having the user accept things twice - after all, they already did accept the path in their file selector, so assuming that they know how to use it, there is no benefit I can see from a second confirmation.

No strong feelings either way, though, this isn't a blocker (other than not accessing the private member directly as mentioned).

> * Allow a setting such that the external picker is triggered directly. But is seems nice to still have the option to for example choose to open the download etc.

This would be good to have, similarly to `content.pdfjs`. Not sure how it would look code-wise though, I suspect it would touch different parts of the code. Might be best to defer it to a separate PR.

AckslD commented:
> Thanks and sorry for the delay! I added a couple of comments.

No problem!

> This would be good to have, similarly to content.pdfjs. Not sure how it would look code-wise though, I suspect it would touch different parts of the code. Might be best to defer it to a separate PR.

Sounds good!

I updated things based on your feedback. I have three questions for now:
1. Is the way I call `accept` correct?
2. There is a cyclic import which needs to be resolved (for now the import is within the function), any thoughts on how to best do this.
3. I started trying to add some tests but I'm a bit unsure how to do this. I tried to follow some of the `downloads.feature` cases but when I run the case I added it times out when waiting for the download prompt. I'm also not fully sure what to assert after running the fileselect. And also not fully sure how to share the `@bdd.when` for setting up the fileselect setting between `editor.feature` and `downloads.feature`.

The-Compiler commented on tests/end2end/features/test_downloads_bdd.py:
```diff
@@ -167,3 +168,34 @@ def fifo_should_be_fifo(tmpdir):
     assert not os.path.isfile(str(download_dir / 'fifo'))
+
+
+# TODO how to share this with test_editor_bdd.py?
```
Move it to `tests/end2end/features/conftest.py` instead.

AckslD replied:
Done :+1:

The-Compiler commented on tests/end2end/features/downloads.feature:
```diff
@@ -684,3 +684,16 @@ Feature: Downloading things from a website.
+        And I setup a fake folder fileselector selecting "tests/end2end/data/backforward/" and writes to a temporary file
+        And I open data/downloads/downloads.html
+        And I run :click-element id download
+        And I wait for the download prompt for "*/download.bin"
```
This doesn't work because the tests end up waiting for:

```
----> Waiting for "Asking question <qutebrowser.utils.usertypes.Question default='*/download.bin' mode=<PromptMode.do… in the log
```

but the log output reads:

```
11:25:14.821 DEBUG    prompt     prompt:ask_question:197 Starting loop.exec() for <qutebrowser.utils.usertypes.Question default='/tmp/pytest-of-florian/pytest-1/test_select_download_path0/downloads/' mode=<PromptMode.download: 5> option=None text='Please enter a location for <b>http://localhost:59185/data/downloads/download.bin</b>' title='Save file to:'>
```

note the `default='*/download.bin'` vs. `default='/tmp/pytest-of-florian/pytest-1/test_select_download_path0/downloads/' - in other words, matching the filename here requires `And I set downloads.location.suggestion to both`.

I'd just do

```suggestion
        And I wait for the download prompt for "*"
```

like in some of the other tests, since the filename doesn't matter for what we actually want to test.

The-Compiler commented on tests/end2end/features/downloads.feature:
```diff
@@ -684,3 +684,16 @@ Feature: Downloading things from a website.
+        And I open data/downloads/downloads.html
+        And I run :click-element id download
+        And I wait for the download prompt for "*/download.bin"
+        And I run :prompt-external-picker
```
```suggestion
        And I run :prompt-fileselect-external
```

The-Compiler commented on tests/end2end/features/downloads.feature:
```diff
@@ -684,3 +684,16 @@ Feature: Downloading things from a website.
+    Scenario: Select download path
+        When I set downloads.location.prompt to true
+        And I set fileselect.handler to external
+        And I setup a fake folder fileselector selecting "tests/end2end/data/backforward/" and writes to a temporary file
```
```suggestion
        And I setup a fake folder fileselector selecting "(tmpdir)(dirsep)downloads(dirsep)subdir" and writes to a temporary file
```

to place the file into a well-defined directory which isn't the default. Search for `(tmpdir)` in this file for some inspiration - the subdir gets created by `Given I set up a temporary download dir` via the `download_dir` fixture in `test_downloads_bdd.py`.

Note that you will need to also adjust the Python code to carry out those replacements, as they are not done universally (thinking about it, maybe they should be?). See `run_command` in `tests/end2end/features/conftest.py` for inspiration.

The-Compiler commented on tests/end2end/features/downloads.feature:
```diff
@@ -684,3 +684,16 @@ Feature: Downloading things from a website.
+        And I wait for the download prompt for "*/download.bin"
+        And I run :prompt-external-picker
+        # TODO what should the check be?
+        Then the javascript message "Files: 1.txt" should be logged
```
With the change above:

```suggestion
        And I wait until the download is finished
        Then the downloaded file subdir/download.bin should exist
```

The-Compiler commented on qutebrowser/mainwindow/prompt.py:
```diff
@@ -467,6 +467,31 @@ def prompt_yank(self, sel=False):
+                "No external location picker (fileselect.folder.command) configured."
+            )
+        # TODO fix cyclic import
+        from qutebrowser.browser import shared
```
The cycle is:

- `browser.browsertab`
- `misc.sessions` (for `sessions.session_manager.save_autosave()` in `_on_load_finished`)
- `mainwindow.mainwindow` (to create new windows from sessions)
- `mainwindow.prompt` (to add the prompt container)
- `browser.shared` (new, for file selection)
- `misc.guiprocess` (to spawn the file selector)
- `api.apitypes` (for `apitypes.Tab` type annotation)
- `browser.browsertab` (`apitypes.Tab` is an alias to `browsertab.Tab`)

No obvious point to break the cycle right now.
This solution is fine for now, I suppose. It's a bigger problem all across the code base, because it's not something I paid attention to when qutebrowser initially grew (I didn't know better at the time). At some point I should probably sit down with a piece of paper and find a suitable split between modules which doesn't require low-level modules to import things from higher-level ones. But that will be a bigger refactoring.

A different solution would be to move `from qutebrowser.misc import guiprocess` to `_execute_fileselect_command` in `browser.shared`. No hard feelings either way, though.

AckslD replied:
Can be tricky with these cyclic imports I find. For now I left the import as is but changed the comment above it.

The-Compiler reviewed (changes requested):
Answered your questions inline.

AckslD commented on qutebrowser/mainwindow/prompt.py:
```diff
@@ -467,6 +467,31 @@ def prompt_yank(self, sel=False):
+        # choose_file already checks that this is max one folder
+        assert len(folders) == 1
+        folder = folders[0]
+        self._prompt.accept(folder)
```
Btw, I changed this to `self.prompt_accept`, does that make sense? The tests didn't work before but with this change they do.

The-Compiler replied:
Ah yes. I thought that `prompt_accept` only called `self._prompt.accept`, but it also does some cleanup which is needed too. :+1:

AckslD commented:
Thanks for the help @The-Compiler! Tests seem to work  now :+1: I'll start adding a few more cases.

AckslD commented:
I added one more scenario where no folder is chosen by the external fileselector. This seems to work but I also wanted to assert that the download prompt is then still open. I tried to do this with
```
        And the download prompt should be shown with "(tmpdir)(dirsep)downloads"
```
but I think this only waits for a new prompt to appear. Is that true? How can one assert that a given prompt is currently shown?

Also I realised that I couldn't come up with many other test cases (especially since we already test a bunch of stuff for the upload fileselector commands). Let me know if you have some cases you think should be added.

The-Compiler commented:
> I added one more scenario where no folder is chosen by the external fileselector. This seems to work but I also wanted to assert that the download prompt is then still open. I tried to do this with
>
> ```
>         And the download prompt should be shown with "(tmpdir)(dirsep)downloads"
> ```
> 
> but I think this only waits for a new prompt to appear. Is that true? How can one assert that a given prompt is currently shown?

Yes, exactly. Those end-to-end tests work by launching qutebrowser in a subprocess, sending commands to it and then observing the logging output and webserver requests.

Thus, it follows that we can't really look at the "current state" of the process, as those are essentially black-box tests. Sometimes there are ways around that (e.g. telling qutebrowser to run `:session-save` and inspecting the file to see what tabs are open), but in this case, I can't really think of anything good.

The next best thing to do here is to check whether the prompt has not been closed - I did that in 435e880477c6573b34464375059d611c9942dd60 by using `And "No prompts left, hiding prompt container." should not be logged`.

> Also I realised that I couldn't come up with many other test cases (especially since we already test a bunch of stuff for the upload fileselector commands). Let me know if you have some cases you think should be added.

The only thing I could come up with is checking for the error when running the command on a non-fileselect prompt. I added such a test in fc17602f79f84d2d2f2ae2946b83f008306ec8a6.

---

> I tried this and it works as advertised. I used ranger, and had to add this to my config (I was confused at the beginning about which option should be set):

This is great feedback! I added some additional explanation to the command help in f4e899e2afface41e05d392bfa5734edb63ca5f1.

---

Thanks @AckslD for the PR and @agenbite for testing! :sparkles:
